### PR TITLE
Add stack of tasks berdy feature 

### PR DESCRIPTION
**Pull request by Yeshasvitvs**
This PR add stack of tasks berdy feature to iDynTree Berdy MAP Estimator

### CODE REVIEW

File: src/estimation/CMakeLists.txt [hunk 1]
```diff
@@ -6,6 +6,8 @@
 # https://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
 # at your option.
 
+project(iDynTree_Estimation CXX)
```
Comment (traversaro) [inline src/estimation/CMakeLists.txt:9]: Is this necessary? Unless a sub-directory is meant to be built on its own by CMake, there is no reason to but `project`  in  a CMakeLists.txt

File: src/estimation/include/iDynTree/Estimation/BerdyHelper.h [hunk 2]
```diff
@@ -689,6 +833,34 @@ class BerdyHelper
...
+    /**
+     * Set the kinematic information necessary for the dynamics estimation using the
+     * angular velocity information of a floating frame.
+     *
+     * \note This method cannot be used if the selected BerdyVariant is ORIGINAL_BERDY_FIXED_BASE.
+     * \note we not require to give the linear velocity of floating base because the dynamics equations
+     *       are invariant with respect to an offset in linear velocity. This convenient to avoid
+     *       any dependency on any prior floating base estimation.
+     *
+     * @param[in] baseTransform the transformation from floating base to world
```
Comment (traversaro) [inline src/estimation/include/iDynTree/Estimation/BerdyHelper.h:850]: What do yo mean "from floating base to world" . I think using a notation like W_H_B or similar is more clear.

File: src/estimation/include/iDynTree/Estimation/BerdyHelper.h [hunk 3]
```diff
@@ -228,6 +256,12 @@ struct BerdyOptions
      */
     std::string baseLink;
 
+    /**
+     * This is a boolean flag for stack of task MAP estimation
+     * The default value is false
+     */
+    bool stackOfTasksMAP;
```
Comment (traversaro) [inline src/estimation/include/iDynTree/Estimation/BerdyHelper.h:263]: Is this used somewhere?

**Review by traversaro — CHANGES_REQUESTED** [review]
I tried to follow the code, but is indeed quite complicate and risk to be problematic to maintain. In particular, boolean arguments should be avoided as much as possible for readability, see https://abseil.io/tips/94 . Also duplicate methods that have the original version without arguments and the new version with just a boolean arguments are particularly problematic. 

I have a small proposal that I think it can dramatically improve the readability of the code: instead of having endless chain of boolean `task1` arguments, can't we just add a new variant to the `BerdyVariants` enum, i.e. on top of: 
* `ORIGINAL_BERDY_FIXED_BASE` 
* `BERDY_FLOATING_BASE`
we can add:
* `HIERARCHICAL_BERDY_FLOATING_BASE`

And just have all the task1/task2 code depend on this variant. What do you think? 

I think to verify that everything works as expected you could add this new variant to the test in https://github.com/robotology/idyntree/blob/84d216c14eb923fb2c54e86e46511565617f314c/src/estimation/tests/BerdyHelperUnitTest.cpp, that basically just check that the matrices are coherent.

File: src/estimation/CMakeLists.txt (hunk 1 above)
Comment (Yeshasvitvs) [reply src/estimation/CMakeLists.txt]: Removed in https://github.com/robotology/idyntree/pull/783/commits/623493079d6aebd58c8fb58938e2141b7289988d

**Comment by Yeshasvitvs** [comment]
> Also duplicate methods that have the original version without arguments and the new version with just a boolean arguments are particularly problematic.

Totally agree on this. 

> I have a small proposal that I think it can dramatically improve the readability of the code: instead of having endless chain of boolean task1 arguments, can't we just add a new variant to the BerdyVariants enum, i.e. on top of:

That makes sense, I will try to restructured accordingly

**Comment by traversaro** [comment]
> That makes sense, I will try to restructure accordingly

Perfect! If you have a pre-print of the paper, you can refer it in the documentation of the variant.

File: src/estimation/include/iDynTree/Estimation/BerdyHelper.h (hunk 3 above)
Comment (Yeshasvitvs) [reply src/estimation/include/iDynTree/Estimation/BerdyHelper.h]: Removed in https://github.com/robotology/idyntree/pull/783/commits/4d0f4da1abc9a7a416414aecad372f0fbcc5e054

**Comment by Yeshasvitvs** [comment]
> I have a small proposal that I think it can dramatically improve the readability of the code: instead of having endless chain of boolean task1 arguments, can't we just add a new variant to the BerdyVariants enum, i.e. on top of:

Following the above suggestions a new Berdy variant `HIERARCHICAL_BERDY_FLOATING_BASE` is added 

> And just have all the task1/task2 code depend on this variant. What do you think?

New definitions of task `CENTROIDAL_DYNAMICS` and `FULL_DYNAMICS` is defined and the original methods are updated with an optional parameter for the task, where `FULL_DYNAMICS` is considered as the default option to ensure the backwards compatibility.

 The above changes are done in https://github.com/robotology/idyntree/pull/783/commits/4d0f4da1abc9a7a416414aecad372f0fbcc5e054 @traversaro please have a look at it and let me know your comments. Thanks.

File: src/estimation/include/iDynTree/Estimation/BerdyHelper.h (hunk 2 above)
Comment (Yeshasvitvs) [reply src/estimation/include/iDynTree/Estimation/BerdyHelper.h:779]: Done in https://github.com/robotology/idyntree/pull/783/commits/bc56f3aeae3daab557e5ac6c946a2baf61dc979e

**Comment by traversaro** [comment]
Hi @Yeshasvitvs , what is the status of this? Last time I saw this comment https://github.com/robotology/idyntree/pull/783#issuecomment-735819475 the tests were not passing, so I did not looked into it. Now I saw that you added more commit for fixing the tests, thanks! However, for the future if you do this, please also comment the PR to notify that the PR is ready for review and tests passes, as repo mantainers don't get notifications for simple commits.

File: src/estimation/include/iDynTree/Estimation/BerdyHelper.h [hunk 4]
```diff
@@ -128,7 +156,13 @@ enum BerdySensorTypes
     /**
      * Non-physical sensor that measures the wrench trasmitted by a joint.
      */
-    JOINT_WRENCH_SENSOR     = 1003
+    JOINT_WRENCH_SENSOR     = 1003,
+
+    /**
+     * Non-physical sensor that holds the value of Rate of Change of Momentum (ROCM) of the system
```
Comment (traversaro) [inline src/estimation/include/iDynTree/Estimation/BerdyHelper.h:162]: Minor, for acronyms you typically capitalize all the letters that are in the acronym
```suggestion
     * Non-physical sensor that holds the value of Rate Of Change of Momentum (ROCM) of the system
```

File: src/estimation/include/iDynTree/Estimation/BerdyHelper.h [hunk 5]
```diff
@@ -677,17 +776,20 @@ class BerdyHelper
      *       are invariant with respect to an offset in linear velocity. This convenient to avoid
      *       any dependency on any prior floating base estimation.
      *
+     * @param[in] baseTransform the transformation from floating base to world
      * @param[in] jointPos the position of the joints of the model.
      * @param[in] jointVel the velocities of the joints of the model.
      * @param[in] floatingFrame the index of the frame for which kinematic information is provided.
      * @param[in] angularVel angular velocity (wrt to any inertial frame) of the specified floating frame,
      *                       expressed in the specified floating frame orientation.
+     * @param[in] w_H_b wptional transformation between floating base and world with default value set to Identity
```
Comment (traversaro) [inline src/estimation/include/iDynTree/Estimation/BerdyHelper.h:785]:
```suggestion
     * @param[in] w_H_b optional transformation between floating base and world with default value set to Identity
```

File: src/estimation/include/iDynTree/Estimation/BerdyHelper.h [hunk 6]
```diff
@@ -677,17 +776,20 @@ class BerdyHelper
      *       are invariant with respect to an offset in linear velocity. This convenient to avoid
      *       any dependency on any prior floating base estimation.
      *
+     * @param[in] baseTransform the transformation from floating base to world
```
Comment (traversaro) [inline src/estimation/include/iDynTree/Estimation/BerdyHelper.h:779]: This line does not seems to correspond to an argument that is actually passed to the function.

File: src/estimation/include/iDynTree/Estimation/BerdySparseMAPSolver.h [hunk 7]
```diff
@@ -68,14 +72,14 @@ namespace iDynTree {
...
                                                    const Vector3& bodyAngularVelocityOfSpecifiedFrame,
-                                                   const VectorDynSize& measurements);
-
-        bool doEstimate();
+                                                   const VectorDynSize& measurements,
+                                                   const Transform& w_H_b = iDynTree::Transform::Identity(),
+                                                   const HierarchialBerdyTask& task = HierarchialBerdyTask::FULL_DYNAMICS);
 
-        void getLastEstimate(iDynTree::VectorDynSize& lastEstimate) const;
+        bool doEstimate(const HierarchialBerdyTask& task = HierarchialBerdyTask::FULL_DYNAMICS);
```
Comment (traversaro) [inline src/estimation/include/iDynTree/Estimation/BerdySparseMAPSolver.h:79]:
```suggestion
        bool doEstimate(const HierarchialBerdyTask task = HierarchialBerdyTask::FULL_DYNAMICS);
```
No need to have references (i.e. pointers) when you just need to pass an enum value (see https://stackoverflow.com/questions/26552481/which-is-faster-pass-by-reference-vs-pass-by-value-c ).

File: src/estimation/include/iDynTree/Estimation/BerdySparseMAPSolver.h [hunk 8]
```diff
@@ -68,14 +72,14 @@ namespace iDynTree {
...
+        void getLastEstimate(iDynTree::VectorDynSize& lastEstimate, const  HierarchialBerdyTask& task  = HierarchialBerdyTask::FULL_DYNAMICS) const;
```
Comment (traversaro) [inline src/estimation/include/iDynTree/Estimation/BerdySparseMAPSolver.h:80]:
```suggestion
        void getLastEstimate(iDynTree::VectorDynSize& lastEstimate, const  HierarchialBerdyTask task  = HierarchialBerdyTask::FULL_DYNAMICS) const;
```

**Comment by traversaro** [comment]
Thanks @Yeshasvitvs . The PR now contains a lot of untested code, I think we can easily add a test (ensuring consistency, i.e. that `D*d + bd = 0` and that `Y*d + by = y` by modifying the test in https://github.com/robotology/idyntree/blob/v2.0.2/src/estimation/tests/BerdyHelperUnitTest.cpp#L348 and adding a few lines to test the new variant, for example with this code:
~~~cxx
    options.berdyVariant = iDynTree::HIERARCHICAL_BERDY_FLOATING_BASE ;
    // Set any option you want to test 
    options.includeAllNetExternalWrenchesAsDynamicVariables = true;
    options.includeAllJointTorquesAsSensors = true;
    ok = berdyHelper.init(estimator.model(), estimator.sensors(), options);
    ASSERT_IS_TRUE(ok);
    testBerdySensorMatrices(berdyHelper, fileName);
~~~

File: src/core/include/iDynTree/Core/EigenHelpers.h [hunk 9]
```diff
@@ -386,6 +386,15 @@ inline void setSubVector(VectorDynSize& vec,
...
 }
 
+inline void setSubVector(VectorDynSize& vec,
+                         const IndexRange range,
+                         const LinearMotionVector3& linMotionVector)
+{
+    assert(range.size==3);
+    toEigen(vec).segment(range.offset, range.size) = toEigen(linMotionVector);
+    return;
+}
```
Comment (traversaro) [inline src/core/include/iDynTree/Core/EigenHelpers.h:396]: Is this necessary? Now `LinearMotionVector3` is just a `GeomVector3`, that inherits from `VectorFixSize<3>, so the previous setSubVector overload should apply also here. Anyhow, if this is necessary I guess we should set it to `GeomVector3`, as `LinearMotionVector3` is deprecated.

**Comment by traversaro** [comment]
As discuss t2t no work is planned at the moment on this PR, so I would close it for now. Feel free to re-open it when you are planning to work on it.

**Pull request closed without merging**